crates/ruff_python_ast/src/nodes.rs

```diff
@@ -1180,8 +1180,48 @@ bitflags! {
         /// it begins and ends with three consecutive quote characters.
         const TRIPLE_QUOTED = 1 << 1;
 
-        /// The f-string has an `r` or `R` prefix, meaning it is a raw f-string.
-        const R_PREFIX = 1 << 3;
+        /// The f-string has an `r` prefix, meaning it is a raw f-string
+        /// with a lowercase 'r'.
+        const R_PREFIX_LOWER = 1 << 2;
+
+        /// The f-string has an `R` prefix, meaning it is a raw f-string
+        /// with an uppercase 'r'.
+        /// See https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#r-strings-and-r-strings
+        /// for why we track the casing of the `r` prefix,
+        /// but not for any other prefix
+        const R_PREFIX_UPPER = 1 << 3;
+    }
+}
+
+/// Enumeration of the valid prefixes an f-string literal can have.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum FStringPrefix {
+    /// Just a regular f-string with no other prefixes
+    Format,
+
+    /// A "raw" format-string, that has an `r` or `R` prefix
+    RawFormat { uppercase_r: bool },
```

Comment:
I'm somewhat torn here. I agree with you that it's easier to type if they're two separate variants. But semantically, they're the same kind of string in Python, just represented in two different ways; having them be two flavours of the same variant feels more expressive of the semantics here. I also think that if I were matching on the prefix in a lint rule (for example), it would be more natural to do this:

```rs
match string_expr.flags.prefix() {
    StringLiteralPrefix::Regular => ...,
    StringLiteralPrefix::Raw { .. } => ...,
}
```

rather than

```rs
match string_expr.flags.prefix() {
    StringLiteralPrefix::Regular => ...,
    StringLiteralPrefix::RawLower | StringLiteralPrefix::RawUpper => ...,
}
```

(In most lint rules, I doubt you'll care much whether it's an `R` or an `r`, since they have the same semantics at runtime for Python.)